// detox/lib/devices/Simulator.js
'use strict';var _createClass = function () {function defineProperties(target, props) {for (var i = 0; i < props.length; i++) {var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);}}return function (Constructor, protoProps, staticProps) {if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;};}();var _get = function get(object, property, receiver) {if (object === null) object = Function.prototype;var desc = Object.getOwnPropertyDescriptor(object, property);if (desc === undefined) {var parent = Object.getPrototypeOf(object);if (parent === null) {return undefined;} else {return get(parent, property, receiver);}} else if ("value" in desc) {return desc.value;} else {var getter = desc.get;if (getter === undefined) {return undefined;}return getter.call(receiver);}};function _asyncToGenerator(fn) {return function () {var gen = fn.apply(this, arguments);return new Promise(function (resolve, reject) {function step(key, arg) {try {var info = gen[key](arg);var value = info.value;} catch (error) {reject(error);return;}if (info.done) {resolve(value);} else {return Promise.resolve(value).then(function (value) {step("next", value);}, function (err) {step("throw", err);});}}return step("next");});};}function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self, call) {if (!self) {throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call && (typeof call === "object" || typeof call === "function") ? call : self;}function _inherits(subClass, superClass) {if (typeof superClass !== "function" && superClass !== null) {throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);}subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } });if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;}var exec = require('child-process-promise').exec;
var path = require('path');
var fs = require('fs');
var os = require('os');
var _ = require('lodash');
var IosNoneDevice = require('./IosNoneDevice');
var FBsimctl = require('./Fbsimctl');
var configuration = require('../configuration');
var argparse = require('../utils/argparse');
var invoke = require('../invoke');var

Simulator = function (_IosNoneDevice) {_inherits(Simulator, _IosNoneDevice);

  function Simulator(client, session, deviceConfig) {_classCallCheck(this, Simulator);var _this = _possibleConstructorReturn(this, (Simulator.__proto__ || Object.getPrototypeOf(Simulator)).call(this,
    client, session, deviceConfig));
    _this._fbsimctl = new FBsimctl();
    _this._simulatorUdid = "";
    _this.sim = "";return _this;
  }_createClass(Simulator, [{ key: 'validate', value: function validate(

    deviceConfig) {
      _get(Simulator.prototype.__proto__ || Object.getPrototypeOf(Simulator.prototype), 'validate', this).call(this, deviceConfig);

      if (!deviceConfig.binaryPath) {
        configuration.throwOnEmptyBinaryPath();
      }

      if (!deviceConfig.name) {
        configuration.throwOnEmptyName();
      }
    } }, { key: '_getBundleIdFromApp', value: function () {var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee(

      appPath) {var absPath, result;return regeneratorRuntime.wrap(function _callee$(_context) {while (1) {switch (_context.prev = _context.next) {case 0:
                absPath = this._getAppAbsolutePath(appPath);_context.prev = 1;_context.next = 4;return (

                  exec('/usr/libexec/PlistBuddy -c "Print CFBundleIdentifier" ' + path.join(absPath, 'Info.plist')));case 4:result = _context.sent;return _context.abrupt('return',
                _.trim(result.stdout));case 8:_context.prev = 8;_context.t0 = _context['catch'](1);throw (

                  new Error('field CFBundleIdentifier not found inside Info.plist of app binary at ' + absPath));case 11:case 'end':return _context.stop();}}}, _callee, this, [[1, 8]]);}));function _getBundleIdFromApp(_x) {return _ref.apply(this, arguments);}return _getBundleIdFromApp;}() }, { key: '_getAppAbsolutePath', value: function _getAppAbsolutePath(



    appPath) {
      var absPath = path.join(process.cwd(), appPath);
      if (fs.existsSync(absPath)) {
        return absPath;
      } else {
        throw new Error('app binary not found at ' + absPath + ', did you build it?');
      }
    } }, { key: 'prepare', value: function () {var _ref2 = _asyncToGenerator(regeneratorRuntime.mark(function _callee2() {return regeneratorRuntime.wrap(function _callee2$(_context2) {while (1) {switch (_context2.prev = _context2.next) {case 0:_context2.next = 2;return (


                  this._fbsimctl.list(this._deviceConfig.name));case 2:this._simulatorUdid = _context2.sent;_context2.next = 5;return (
                  this._getBundleIdFromApp(this._deviceConfig.binaryPath));case 5:this._bundleId = _context2.sent;_context2.next = 8;return (
                  this._fbsimctl.boot(this._simulatorUdid));case 8:_context2.next = 10;return (
                  this.relaunchApp({ delete: !argparse.getArgValue('reuse') }));case 10:case 'end':return _context2.stop();}}}, _callee2, this);}));function prepare() {return _ref2.apply(this, arguments);}return prepare;}() }, { key: 'relaunchApp', value: function () {var _ref3 = _asyncToGenerator(regeneratorRuntime.mark(function _callee3() {var


        params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};var bundleId = arguments[1];var additionalLaunchArgs, _bundleId;return regeneratorRuntime.wrap(function _callee3$(_context3) {while (1) {switch (_context3.prev = _context3.next) {case 0:if (!(
                params.url && params.userNotification)) {_context3.next = 2;break;}throw (
                  new Error('detox can\'t understand this \'relaunchApp(' + JSON.stringify(params) + ')\' request, either request to launch with url or with userNotification, not both'));case 2:if (!


                params.delete) {_context3.next = 9;break;}_context3.next = 5;return (
                  this._fbsimctl.uninstall(this._simulatorUdid, this._bundleId));case 5:_context3.next = 7;return (
                  this._fbsimctl.install(this._simulatorUdid, this._getAppAbsolutePath(this._deviceConfig.binaryPath)));case 7:_context3.next = 11;break;case 9:_context3.next = 11;return (


                  this._fbsimctl.terminate(this._simulatorUdid, this._bundleId));case 11:


                additionalLaunchArgs = void 0;
                if (params.url) {
                  additionalLaunchArgs = { '-detoxURLOverride': params.url };
                } else if (params.userNotification) {
                  additionalLaunchArgs = { '-detoxUserNotificationDataURL': this.createPushNotificationJson(params.userNotification) };
                }

                _bundleId = bundleId || this._bundleId;_context3.next = 16;return (
                  this._fbsimctl.launch(this._simulatorUdid, _bundleId, this._prepareLaunchArgs(additionalLaunchArgs)));case 16:_context3.next = 18;return (
                  this.client.waitUntilReady());case 18:case 'end':return _context3.stop();}}}, _callee3, this);}));function relaunchApp() {return _ref3.apply(this, arguments);}return relaunchApp;}() }, { key: 'installApp', value: function () {var _ref4 = _asyncToGenerator(regeneratorRuntime.mark(function _callee4(


      binaryPath) {var _binaryPath;return regeneratorRuntime.wrap(function _callee4$(_context4) {while (1) {switch (_context4.prev = _context4.next) {case 0:
                _binaryPath = binaryPath || this._getAppAbsolutePath(this._deviceConfig.binaryPath);_context4.next = 3;return (
                  this._fbsimctl.install(this._simulatorUdid, _binaryPath));case 3:case 'end':return _context4.stop();}}}, _callee4, this);}));function installApp(_x3) {return _ref4.apply(this, arguments);}return installApp;}() }, { key: 'uninstallApp', value: function () {var _ref5 = _asyncToGenerator(regeneratorRuntime.mark(function _callee5(


      bundleId) {var _bundleId;return regeneratorRuntime.wrap(function _callee5$(_context5) {while (1) {switch (_context5.prev = _context5.next) {case 0:
                _bundleId = bundleId || this._bundleId;_context5.next = 3;return (
                  this._fbsimctl.uninstall(this._simulatorUdid, _bundleId));case 3:case 'end':return _context5.stop();}}}, _callee5, this);}));function uninstallApp(_x4) {return _ref5.apply(this, arguments);}return uninstallApp;}() }, { key: 'openURL', value: function () {var _ref6 = _asyncToGenerator(regeneratorRuntime.mark(function _callee6(


      url) {return regeneratorRuntime.wrap(function _callee6$(_context6) {while (1) {switch (_context6.prev = _context6.next) {case 0:_context6.next = 2;return (
                  this._fbsimctl.open(this._simulatorUdid, url));case 2:case 'end':return _context6.stop();}}}, _callee6, this);}));function openURL(_x5) {return _ref6.apply(this, arguments);}return openURL;}() }, { key: 'shutdown', value: function () {var _ref7 = _asyncToGenerator(regeneratorRuntime.mark(function _callee7() {return regeneratorRuntime.wrap(function _callee7$(_context7) {while (1) {switch (_context7.prev = _context7.next) {case 0:_context7.next = 2;return (



                  this._fbsimctl.shutdown(this._simulatorUdid));case 2:case 'end':return _context7.stop();}}}, _callee7, this);}));function shutdown() {return _ref7.apply(this, arguments);}return shutdown;}() }, { key: 'setOrientation', value: function () {var _ref8 = _asyncToGenerator(regeneratorRuntime.mark(function _callee8(


      orientation) {var orientationMapping, call;return regeneratorRuntime.wrap(function _callee8$(_context8) {while (1) {switch (_context8.prev = _context8.next) {case 0:
                // keys are possible orientations
                orientationMapping = {
                  landscape: 3, // top at left side landscape
                  portrait: 1 // non-reversed portrait
                };if (
                Object.keys(orientationMapping).includes(orientation)) {_context8.next = 3;break;}throw (
                  new Error('setOrientation failed: provided orientation ' + orientation + ' is not part of supported orientations: ' + Object.keys(orientationMapping)));case 3:


                call = invoke.call(invoke.EarlGrey.instance,
                'rotateDeviceToOrientation:errorOrNil:',
                invoke.IOS.NSInteger(orientationMapping[orientation]));_context8.next = 6;return (

                  new invoke.InvocationManager(this.client).execute(call));case 6:case 'end':return _context8.stop();}}}, _callee8, this);}));function setOrientation(_x6) {return _ref8.apply(this, arguments);}return setOrientation;}() }]);return Simulator;}(IosNoneDevice);



module.exports = Simulator;
