// detox/lib/configurations.mock.js
"use strict";var validOneDeviceNoSession = {
  "configurations": {
    "ios.sim.release": {
      "binaryPath": "ios/build/Build/Products/Release-iphonesimulator/example.app",
      "type": "ios.simulator",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var validOneIosNoneDeviceNoSession = {
  "configurations": {
    "ios.sim.none": {
      "binaryPath": "ios/build/Build/Products/Release-iphonesimulator/example.app",
      "type": "ios.none",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var validTwoDevicesNoSession = {
  "configurations": {
    "ios.sim.release": {
      "binaryPath": "ios/build/Build/Products/Release-iphonesimulator/example.app",
      "type": "ios.simulator",
      "name": "iPhone 7 Plus, iOS 10.2" },

    "ios.sim.debug": {
      "binaryPath": "ios/build/Build/Products/Debug-iphonesimulator/example.app",
      "type": "ios.simulator",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var invalidDeviceNoBinary = {
  "configurations": {
    "ios.sim.release": {
      "type": "ios.simulator",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var invalidNoDevice = {
  "configurations": {} };



var invalidDeviceNoDeviceType = {
  "configurations": {
    "ios.sim.release": {
      "binaryPath": "here",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var invalidDeviceNoDeviceName = {
  "configurations": {
    "ios.sim.release": {
      "binaryPath": "here",
      "type": "ios.simulator" } } };




var validOneDeviceAndSession = {
  "session": {
    "server": "ws://localhost:8099",
    "sessionId": "test" },

  "configurations": {
    "ios.sim.release": {
      "binaryPath": "ios/build/Build/Products/Release-iphonesimulator/example.app",
      "type": "ios.simulator",
      "name": "iPhone 7 Plus, iOS 10.2" } } };




var invalidSessionNoSessionId = {
  "session": {
    "server": "ws://localhost:8099" } };



var invalidSessionNoServer = {
  "session": {
    "sessionId": "test" } };



var invalidOneDeviceTypeEmulatorNoSession = {
  "configurations": {
    "ios.sim.release": {
      "binaryPath": "some.apk",
      "type": "emulator",
      "name": "an emulator" } } };




module.exports = {
  validOneDeviceNoSession: validOneDeviceNoSession,
  validOneIosNoneDeviceNoSession: validOneIosNoneDeviceNoSession,
  validTwoDevicesNoSession: validTwoDevicesNoSession,
  invalidDeviceNoBinary: invalidDeviceNoBinary,
  invalidNoDevice: invalidNoDevice,
  invalidDeviceNoDeviceType: invalidDeviceNoDeviceType,
  invalidDeviceNoDeviceName: invalidDeviceNoDeviceName,
  validOneDeviceAndSession: validOneDeviceAndSession,
  invalidSessionNoSessionId: invalidSessionNoSessionId,
  invalidSessionNoServer: invalidSessionNoServer,
  invalidOneDeviceTypeEmulatorNoSession: invalidOneDeviceTypeEmulatorNoSession };
