// detox/lib/ios/matchers.js
'use strict';var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {return typeof obj;} : function (obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};var _createClass = function () {function defineProperties(target, props) {for (var i = 0; i < props.length; i++) {var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);}}return function (Constructor, protoProps, staticProps) {if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;};}();function _possibleConstructorReturn(self, call) {if (!self) {throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call && (typeof call === "object" || typeof call === "function") ? call : self;}function _inherits(subClass, superClass) {if (typeof superClass !== "function" && superClass !== null) {throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);}subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } });if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;}function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}var invoke = require('../invoke');var

Matcher = function () {function Matcher() {_classCallCheck(this, Matcher);}_createClass(Matcher, [{ key: 'withAncestor', value: function withAncestor(
    matcher) {
      if (!(matcher instanceof Matcher)) throw new Error('Matcher withAncestor argument must be a valid Matcher, got ' + (typeof matcher === 'undefined' ? 'undefined' : _typeof(matcher)));
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForBoth:andAncestorMatcher:', _originalMatcherCall, matcher._call);
      return this;
    } }, { key: 'withDescendant', value: function withDescendant(
    matcher) {
      if (!(matcher instanceof Matcher)) throw new Error('Matcher withDescendant argument must be a valid Matcher, got ' + (typeof matcher === 'undefined' ? 'undefined' : _typeof(matcher)));
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForBoth:andDescendantMatcher:', _originalMatcherCall, matcher._call);
      return this;
    } }, { key: 'and', value: function and(
    matcher) {
      if (!(matcher instanceof Matcher)) throw new Error('Matcher and argument must be a valid Matcher, got ' + (typeof matcher === 'undefined' ? 'undefined' : _typeof(matcher)));
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForBoth:and:', _originalMatcherCall, matcher._call);
      return this;
    } }, { key: 'not', value: function not()
    {
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForNot:', _originalMatcherCall);
      return this;
    } }, { key: '_avoidProblematicReactNativeElements', value: function _avoidProblematicReactNativeElements()
    {
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherAvoidingProblematicReactNativeElements:', _originalMatcherCall);
      return this;
    } }, { key: '_extendToDescendantScrollViews', value: function _extendToDescendantScrollViews()
    {
      var _originalMatcherCall = this._call;
      this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForScrollChildOfMatcher:', _originalMatcherCall);
      return this;
    } }]);return Matcher;}();var


LabelMatcher = function (_Matcher) {_inherits(LabelMatcher, _Matcher);
  function LabelMatcher(value) {_classCallCheck(this, LabelMatcher);var _this = _possibleConstructorReturn(this, (LabelMatcher.__proto__ || Object.getPrototypeOf(LabelMatcher)).call(this));

    if (typeof value !== 'string') throw new Error('LabelMatcher ctor argument must be a string, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    _this._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detox_matcherForAccessibilityLabel:', value);return _this;
  }return LabelMatcher;}(Matcher);var


IdMatcher = function (_Matcher2) {_inherits(IdMatcher, _Matcher2);
  function IdMatcher(value) {_classCallCheck(this, IdMatcher);var _this2 = _possibleConstructorReturn(this, (IdMatcher.__proto__ || Object.getPrototypeOf(IdMatcher)).call(this));

    if (typeof value !== 'string') throw new Error('IdMatcher ctor argument must be a string, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    _this2._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForAccessibilityID:', value);return _this2;
  }return IdMatcher;}(Matcher);var


TypeMatcher = function (_Matcher3) {_inherits(TypeMatcher, _Matcher3);
  function TypeMatcher(value) {_classCallCheck(this, TypeMatcher);var _this3 = _possibleConstructorReturn(this, (TypeMatcher.__proto__ || Object.getPrototypeOf(TypeMatcher)).call(this));

    if (typeof value !== 'string') throw new Error('TypeMatcher ctor argument must be a string, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    _this3._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForClass:', value);return _this3;
  }return TypeMatcher;}(Matcher);var


TraitsMatcher = function (_Matcher4) {_inherits(TraitsMatcher, _Matcher4);
  function TraitsMatcher(value) {_classCallCheck(this, TraitsMatcher);var _this4 = _possibleConstructorReturn(this, (TraitsMatcher.__proto__ || Object.getPrototypeOf(TraitsMatcher)).call(this));

    if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) !== 'object' || !value instanceof Array) throw new Error('TraitsMatcher ctor argument must be an array, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    var traits = 0;
    for (var i = 0; i < value.length; i++) {
      switch (value[i]) {
        case 'button':traits |= 1;break;
        case 'link':traits |= 2;break;
        case 'header':traits |= 4;break;
        case 'search':traits |= 8;break;
        case 'image':traits |= 16;break;
        case 'selected':traits |= 32;break;
        case 'plays':traits |= 64;break;
        case 'key':traits |= 128;break;
        case 'text':traits |= 256;break;
        case 'summary':traits |= 512;break;
        case 'disabled':traits |= 1024;break;
        case 'frequentUpdates':traits |= 2048;break;
        case 'startsMedia':traits |= 4096;break;
        case 'adjustable':traits |= 8192;break;
        case 'allowsDirectInteraction':traits |= 16384;break;
        case 'pageTurn':traits |= 32768;break;
        default:throw new Error('Unknown trait \'' + value[i] + '\', see list in https://facebook.github.io/react-native/docs/accessibility.html#accessibilitytraits-ios');}

    }
    _this4._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForAccessibilityTraits:', invoke.IOS.NSInteger(traits));return _this4;
  }return TraitsMatcher;}(Matcher);var


VisibleMatcher = function (_Matcher5) {_inherits(VisibleMatcher, _Matcher5);
  function VisibleMatcher() {_classCallCheck(this, VisibleMatcher);var _this5 = _possibleConstructorReturn(this, (VisibleMatcher.__proto__ || Object.getPrototypeOf(VisibleMatcher)).call(this));

    _this5._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForSufficientlyVisible');return _this5;
  }return VisibleMatcher;}(Matcher);var


NotVisibleMatcher = function (_Matcher6) {_inherits(NotVisibleMatcher, _Matcher6);
  function NotVisibleMatcher() {_classCallCheck(this, NotVisibleMatcher);var _this6 = _possibleConstructorReturn(this, (NotVisibleMatcher.__proto__ || Object.getPrototypeOf(NotVisibleMatcher)).call(this));

    _this6._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForNotVisible');return _this6;
  }return NotVisibleMatcher;}(Matcher);var


ExistsMatcher = function (_Matcher7) {_inherits(ExistsMatcher, _Matcher7);
  function ExistsMatcher() {_classCallCheck(this, ExistsMatcher);var _this7 = _possibleConstructorReturn(this, (ExistsMatcher.__proto__ || Object.getPrototypeOf(ExistsMatcher)).call(this));

    _this7._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForNotNil');return _this7;
  }return ExistsMatcher;}(Matcher);var


NotExistsMatcher = function (_Matcher8) {_inherits(NotExistsMatcher, _Matcher8);
  function NotExistsMatcher() {_classCallCheck(this, NotExistsMatcher);var _this8 = _possibleConstructorReturn(this, (NotExistsMatcher.__proto__ || Object.getPrototypeOf(NotExistsMatcher)).call(this));

    _this8._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForNil');return _this8;
  }return NotExistsMatcher;}(Matcher);var


TextMatcher = function (_Matcher9) {_inherits(TextMatcher, _Matcher9);
  function TextMatcher(value) {_classCallCheck(this, TextMatcher);var _this9 = _possibleConstructorReturn(this, (TextMatcher.__proto__ || Object.getPrototypeOf(TextMatcher)).call(this));

    if (typeof value !== 'string') throw new Error('TextMatcher ctor argument must be a string, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    _this9._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'detoxMatcherForText:', value);return _this9;
  }return TextMatcher;}(Matcher);var


ValueMatcher = function (_Matcher10) {_inherits(ValueMatcher, _Matcher10);
  function ValueMatcher(value) {_classCallCheck(this, ValueMatcher);var _this10 = _possibleConstructorReturn(this, (ValueMatcher.__proto__ || Object.getPrototypeOf(ValueMatcher)).call(this));

    if (typeof value !== 'string') throw new Error('ValueMatcher ctor argument must be a string, got ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)));
    _this10._call = invoke.call(invoke.IOS.Class('GREYMatchers'), 'matcherForAccessibilityValue:', value);return _this10;
  }return ValueMatcher;}(Matcher);


module.exports = {
  Matcher: Matcher,
  LabelMatcher: LabelMatcher,
  IdMatcher: IdMatcher,
  TypeMatcher: TypeMatcher,
  TraitsMatcher: TraitsMatcher,
  VisibleMatcher: VisibleMatcher,
  NotVisibleMatcher: NotVisibleMatcher,
  ExistsMatcher: ExistsMatcher,
  NotExistsMatcher: NotExistsMatcher,
  TextMatcher: TextMatcher,
  ValueMatcher: ValueMatcher };
